### PR TITLE
variety of optimizations and additions

### PR DESCRIPTION
**Pull request by patrickhop**
(1) Added a featurize function for numpy arrays loaded with smiles

(2) n_tasks was added to the init function header of NumpyDataset. When conducting inference on a numpy dataset with y=None, the shape was wrong for multitask. This would crash tensorflow which was expecting the shape defined in training.

(3) Added r2_score using coefficients of determination, which is a nice alternative to the square of pearson r, which we use across the board. Particularly, the metric can be negative if your validation predictions are worse than simply the mean of validation predictions.

(4) For tensorgraph graph conv, I added a missing relu activation for the fully connected layer leading into the readout layer. This slightly improves performance across a variety of datasets, and is consistent with the old implementation,

(5) Added predict_on_smiles and predict_on_smiles_batch for easy inference on smile strings for tensorgraph graph convolutions. This will be helpful for when the query smiles are not stored in csv.

Please let me know if there are any questions or concerns about the changes.

### CODE REVIEW

File: deepchem/metrics/__init__.py [hunk 1]
```diff
@@ -71,6 +71,9 @@ def pearson_r2_score(y, y_pred):
   """Computes Pearson R^2 (square of Pearson correlation)."""
   return pearsonr(y, y_pred)[0]**2
 
+def r2_score(y, y_pred):
```
Comment (rbharath) [inline deepchem/metrics/__init__.py]: This function isn't needed (`r2_score` is exported directly from scikit-learn and can be used already)

File: deepchem/models/tensorgraph/models/graph_models.py [hunk 2]
```diff
@@ -712,3 +716,29 @@ def evaluate(self, dataset, metrics, transformers=[], per_task_metrics=False):
         metrics,
         labels=self.my_labels,
         weights=[self.my_task_weights])
+
+  def predict_on_smiles(self, smiles, transformers):
```
Comment (rbharath) [inline deepchem/models/tensorgraph/models/graph_models.py]: These two seem like useful helper methods. +1

**Comment by rbharath** [comment]
@patrickhop Thanks for the fixes! Changes look good overall, except for the `r2_score` definition. This is causing a test failure due to infinite recursion. Could you remove the function?

Also, would you mind adding a unit test or two for the new functions? Good to make sure new features are represented in the test suite.

I like the addition of `predict_on_smiles`. I think it's a nice convenience for users to get predictions out of models easily. Might even be worth moving up to a superclass in a future PR.

File: deepchem/metrics/__init__.py (hunk 1 above)
Comment (patrickhop) [inline deepchem/metrics/__init__.py]: good call. will update

File: deepchem/data/datasets.py [hunk 3]
```diff
@@ -256,9 +256,8 @@ def __init__(self, X, y=None, w=None, ids=None):
           w = np.reshape(w, (n_samples, -1))
       else:
         # Set labels to be zero, with zero weights
-        y = np.zeros((n_samples, 1))
+        y = np.zeros((n_samples, 1*n_tasks))
```
Comment (lilleswing) [inline deepchem/data/datasets.py]: why 1 * n_tasks instead of n_tasks?

Comment (patrickhop) [inline deepchem/data/datasets.py]: fixed.

**Comment by patrickhop** [comment]
@rbharath do i need to fix something to keep travis-ci happy?

**Comment by lilleswing** [comment]
@patrickhop the python3 one is broken because of an implicit float/integer caste
```bash
======================================================================
ERROR: deepchem.models.tensorgraph.models.test_graph_models.test_graph_conv_regression_model
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/anaconda/envs/deepchem/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/travis/build/deepchem/deepchem/deepchem/models/tensorgraph/models/test_graph_models.py", line 62, in test_graph_conv_regression_model
    out = model.predict_on_smiles(mols, transformers)
  File "/home/travis/build/deepchem/deepchem/deepchem/models/tensorgraph/models/graph_models.py", line 725, in predict_on_smiles
    for i in range(num_batches):
TypeError: 'float' object cannot be interpreted as an integer
```

you also have to run yapf
```bash
yapf -i deepchem/data/data_loader.py
deepchem/data/datasets.py
deepchem/metrics/__init__.py
deepchem/models/tensorgraph/models/graph_models.py
deepchem/models/tensorgraph/models/test_graph_models.py
```

**Comment by lilleswing** [comment]
yapf issues -- is this non-determinism or did you not run it?

I have a branch with an updated yapf here which might help if it is non-determinism.
https://github.com/deepchem/deepchem/pull/662

**Comment by rbharath** [comment]
@patrickhop Almost there! Just need to resolve these yapf issues and we can get this merged.

**Comment by patrickhop** [comment]
@lilleswing @rbharath I re-ran yapf recursively on all the folders I touched. Hopefully this does the trick.

**Comment by rbharath** [comment]
@patrickhop Everything looks good now! Thanks for the contribution.

LGTM

**Pull request merged**